Integration file methods

cmcantalupo opened the pull request:
- Add methods to support file management during integration experiments.
- Modify archive cache use to make symbolic links rather than copying archives.

cmcantalupo committed:
Make a link to archive if found in cache, do not copy

Signed-off-by: Christopher M. Cantalupo <christopher.m.cantalupo@intel.com>

cmcantalupo committed:
Add methods to AppConf to support file managment

- New methods enable files to be created and deleted
  to support running an experiment.

Signed-off-by: Christopher M. Cantalupo <christopher.m.cantalupo@intel.com>

cmcantalupo commented on integration/apps/apps.py:
```diff
@@ -137,7 +137,69 @@ def get_cpu_per_rank(self):
+        pass
+
     def get_bash_setup_commands(self):
+
```
Yuck

dianarg commented on integration/apps/apps.py:
```diff
@@ -137,7 +137,69 @@ def get_cpu_per_rank(self):
+
+    def prelaunch_setup(self, run_id, output_dir):
+        '''Execute any commands that will create files to be used during a
+           launch of an step in an experiment.  This may include
```
a step

dianarg commented on integration/apps/apps.py:
```diff
@@ -137,7 +137,69 @@ def get_cpu_per_rank(self):
+           launch of an step in an experiment.  This may include
+           creating symbolic links, copying files into the output
+           directory, or executing commands that create input data for
+           the application.  This step will be called repeately, once
```
repeatedly

dianarg commented on integration/apps/apps.py:
```diff
@@ -137,7 +137,69 @@ def get_cpu_per_rank(self):
+    def postlaunch_teardown(self, run_id, output_dir):
+        '''Delete files or symbolic links created as part of an experiment
+           trial that are not needed by subsequent trials.  This step
+           will be called repeately, once for each trial executed by
```
repeatedly

dianarg commented on integration/apps/apps.py:
```diff
@@ -137,7 +137,69 @@ def get_cpu_per_rank(self):
+        '''
+        pass
+
+    def experiment_teardown(self, output_dir):
```
Can we have an experiment_setup() too?

dianarg commented on integration/experiment/launch_util.py:
```diff
@@ -141,13 +141,16 @@ def launch_all_runs(targets, num_nodes, iterations, extra_cli_args, output_dir,
 
             # rest to cool off between runs
             time.sleep(cool_off_time)
+    app_conf.experiment_teardown(output_dir)
```
this only happens for the last app_conf; loop through all of them?

cmcantalupo committed:
Add method to AppConf to set up experiment and clean up documentation of new methods

Signed-off-by: Christopher M. Cantalupo <christopher.m.cantalupo@intel.com>

cmcantalupo committed:
Fixup missing pass

dianarg commented on integration/apps/apps.py:
```diff
@@ -137,6 +137,79 @@ def get_cpu_per_rank(self):
+           create symbolic links, copy files into the output
+           directory, or execute commands that create input data for
+           the application.
+
```
docstring for output_dir

dianarg commented on integration/apps/apps.py:
```diff
@@ -137,6 +137,79 @@ def get_cpu_per_rank(self):
+    def experiment_setup(self, output_dir):
+        '''Create files to be used throughout an experiment and present for
+           all trials.  These files typically persist unchanged
+           throughout the execution of the experiement and are cleaned
```
experiment, more below

cmcantalupo committed:
Final round of API clean up

- Removed experiment_setup()
- Run experiment_teardown() multiple times.
- Documentation clean up.

Signed-off-by: Christopher M. Cantalupo <christopher.m.cantalupo@intel.com>

cmcantalupo committed:
Use setup and teardown methods in nekbone app conf

Signed-off-by: Christopher M. Cantalupo <christopher.m.cantalupo@intel.com>

bgeltz commented on integration/apps/nekbone/nekbone.py:
```diff
@@ -54,25 +54,32 @@ def __init__(self, add_barriers=False):
+    def trial_setup(self, run_id, output_dir):
         size = 10000  # this size varies per system
-        input_file = textwrap.dedent('''
+        input_data = textwrap.dedent('''
```
The data.rea should always be the same for every trial.

Doesn't this belong in experiment_setup?

cmcantalupo replied:
We got rid of experiment_setup.  The issue is that there are N app_conf objects, but only one start to the experiment.  Instead, the implementer of the trial_setup() is responsible for checking for file existence.
